autonat.go

```diff
@@ -135,50 +135,78 @@ func (as *AmbientAutoNAT) autodetect() {
 	}
 
 	cli := NewAutoNATClient(as.host, as.getAddrs)
-	failures := 0
-
-	for _, pi := range peers {
-		ctx, cancel := context.WithTimeout(as.ctx, AutoNATRequestTimeout)
-		as.host.Peerstore().AddAddrs(pi.ID, pi.Addrs, pstore.TempAddrTTL)
-		a, err := cli.DialBack(ctx, pi.ID)
-		cancel()
-
-		switch {
-		case err == nil:
-			log.Debugf("NAT status is public; address through %s: %s", pi.ID.Pretty(), a.String())
-			as.mx.Lock()
-			as.addr = a
-			as.status = NATStatusPublic
-			as.confidence = 0
-			as.mx.Unlock()
-			return
+	ctx, cancel := context.WithTimeout(as.ctx, AutoNATRequestTimeout)
+	defer cancel()
+
+	var mx sync.Mutex
+	var pubaddr ma.Multiaddr
+	private := 0
```

Comment:
Can we enclose these vars in an anonymous struct? It'll be easier to track the scope of these things, and the zero values are our friend here:

```go
var result struct {
    sync.Mutex
    private int
    public int
    pubaddr ma.Multiaddr
}
```

Then below:

```go
result.Lock()
result.private++
result.Unlock()+
```

Reply:
sure, that's nicer indeed.

Reply:
done